### DISCUSSION TITLE
`git test`

### DISCUSSION THREAD

**Discussion by arxanas** [arxanas/git-branchless#643] [Show and tell]
In v0.6.0-rc.1, `git test` is a sort of replacement for pre-commit hooks. You can use it to run checks on an arbitrary set of commits. For example, in this video, I run the Rust formatter to verify that all commits in the current stack are correctly formatted:

https://user-images.githubusercontent.com/454057/201497775-93895fe6-df3c-41d3-9bb5-5fa7c0ed587b.mov

I find this useful in patch-stack workflows, where I want to make sure that all commits in the stack pass formatting/linting/testing. You can find the documentation at [Command: git test](https://github.com/arxanas/git-branchless/wiki/Command:-git-test). I also find it useful if I want to run a check without switching away from the current commit, as `git test` knows how to create a worktree to run the test with. It also knows how to run commands in parallel, which is useful for low-concurrency or I/O-bound checks.

Currently not implemented:

- Search (neither linear nor binary).
  - Interactive testing is also not implemented.
- Fixing (applying the results of a formatter or linter to the commit in-place).
- Test case minimization.

**Comment by arxanas** [comment]
Here's an example of caching in practice. Notice that we still cache the results of the last commit in the stack because results are cached by tree hash rather than commit hash.

https://user-images.githubusercontent.com/454057/201507456-3bfde946-f3d7-4866-977b-8900a6a18232.mov

**Reply by Gelio** [reply]
What a cool idea!

Minor: consider uploading mp4 files instead of mov, as mov are not universally supported. Here's how it looks for me on Firefox 106.0.5 on Ubuntu 20.04 LTS

![image](https://user-images.githubusercontent.com/889383/201859415-112b89d7-25bf-4bc2-a03f-fd4f73d87be3.png)

It works in Chrome, though.

**Reply by arxanas** [reply]
@Gelio thanks for the note, I'll try to upload mp4s next time!
